text/0000-unicode-string.md

```diff
@@ -0,0 +1,97 @@
+
+# Motivation
+
+Unicode is an accepted standard for representing text in all languages used on planet earth. Modern programming languages should have first class support for unicode.
+
+# Detailed design
+
+Change the builtin String class to present of sequence of unicode codepoints as 4 byte numbers (U32). 
+```
+class val String is (Seq[U32] & Comparable[String box] & Stringable)
+```
+Where String functions took parameters of type U8 or returned values of U8, they will take or return values of type U32. All indexes will be interpreted as codepoints. Only functions that manage String allocated memory will continue to refer to bytes.
+
+The following additional changes will be made to the String class:
+1. The size() function will return the number of unicode codepoints in the string. A new byte_size() function will return the number of bytes.
+1. The truncate() function will only support len parameter values less than the string size.
+1. The utf32() function will be removed. It is redundant, and returns pair that includes a byte count that is no longer needed.
+1. The insert_byte() function will be changed to insert_utf32()
+1. The values() function will return an iterator over the string codepoints. Same as runes().
+1. A concat_bytes() function will be added to add a sequence of codepoints to the string from an iterator of bytes.
+
+Add traits Encoder and Decoder to the builtin package. Any function that produces a String from bytes, or produces bytes from a String must take an Encoder or Decoder as a parameter as is appropriate. 
+```
+trait val Encoder
+  fun encode(codepoint: U32): (USize, U8, U8, U8, U8)
+  
+trait val Decoder
+  fun decode(bytes: U32): (U32, U8)
+```
+
+The ByteSeq type defined in std_stream.pony will be changed to remove String.
+```
+type ByteSeq is (Array[U8] val)
+```
+Many functions that accept ByteSeq as a parameter will be changed to accept (String | ByteSeq) as a parameter.
+
+A new StringIter interface will be added to std_stream.pony
+```
+interface val StringIter
+  """
+  An iterable collection of String box.
+  """
+  fun values(): Iterator[this->String box]
+```
+
+Change Reader in buffered/reader.pony to add functions to read a codepoint and to read a String of a given number of codepoints. Update function line() to accept a decoder, and to return a pair with the line string, and the number of bytes consumed.
+
+Change Writer in buffered/writer.pony to accept Encoder parameters in the write() and writev() functions. 
+
+Add a FileCharacters class in buffered/file_characters.pony that provides an iterator of characters in a file. The implementation will be similar to the FileLines class.
```

Comment:
I think you meant to put this in the `files` package (where it can live alongside `FileLines`) rather than the `buffered` package.

```suggestion
Add a FileCharacters class in files/file_characters.pony that provides an iterator of characters in a file. The implementation will be similar to the FileLines class.
```